src/zasm.zig

```diff
@@ -0,0 +1,716 @@
+// MIT License
+//
+// Copyright (c) 2020-2022 Lee Cannon
+
+const std = @import("std");
+
+// ~~~ [ zig-x86_64/src/registers/control.zig ] ~~~
+
+/// Various control flags modifying the basic operation of the CPU.
+pub const Cr0 = packed struct {
+    /// Enables protected mode.
+    protected_mode: bool,
+
+    /// Enables monitoring of the coprocessor, typical for x87 instructions.
+    ///
+    /// Controls together with the `task_switched` flag whether a `wait` or `fwait`
+    /// instruction should cause an `#NE` exception.
+    monitor_coprocessor: bool,
+
+    /// Force all x87 and MMX instructions to cause an `#NE` exception.
+    emulate_coprocessor: bool,
+
+    /// Automatically set to 1 on _hardware_ task switch.
+    ///
+    /// This flags allows lazily saving x87/MMX/SSE instructions on hardware context switches.
+    task_switched: bool,
+
+    /// Indicates support of 387DX math coprocessor instructions.
+    ///
+    /// Always set on all recent x86 processors, cannot be cleared.
+    extension_type: bool,
+
+    /// Enables the native (internal) error reporting mechanism for x87 FPU errors.
+    numeric_error: bool,
+
+    z_reserved6_15: u10,
+
+    /// Controls whether supervisor-level writes to read-only pages are inhibited.
+    ///
+    /// When set, it is not possible to write to read-only pages from ring 0.
+    write_protect: bool,
+
+    z_reserved17: bool,
+
+    /// Enables automatic usermode alignment checking if `RFlags.alignment_check` is also set.
+    alignment_mask: bool,
+
+    z_reserved19_28: u10,
+
+    /// Ignored. Used to control write-back/write-through cache strategy on older CPUs.
+    not_write_through: bool,
+
+    /// Disables some processor caches, specifics are model-dependent.
+    cache_disable: bool,
+
+    /// Enables paging.
+    ///
+    /// If this bit is set, `protected_mode` must be set.
+    paging: bool,
+
+    z_reserved32_63: u32,
+
+    /// Read the current set of CR0 flags.
+    pub fn read() Cr0 {
+        return Cr0.fromU64(readRaw());
+    }
+
+    /// Read the current raw CR0 value.
+    fn readRaw() u64 {
+        return asm ("mov %%cr0, %[ret]"
+            : [ret] "=r" (-> u64),
+        );
```

Comment:
I read it as `ret` is stored in the return "variable". That is, the function returns a `u64` and the compiler will place the result in e.g. `rax` so, there is no need for us to create a temporary `var ret: u64` local variable, just to have that become the return value.

The `(-> u64)` syntax allows us to do this.

Exactly why that specific syntax was chosen I have no idea :P

Reply:
Yeah, the syntax in your rewrite makes more sense for me. Inline assembly syntax I guess :shrug:

Reply:
> Yeah, the syntax in your rewrite makes more sense for me. Inline assembly syntax I guess shrug

It seems they want to use Intel syntax, but cannot due to LLVM.

ref: https://ziglang.org/documentation/master/#Assembly

>  For i386 and x86_64 targets, the syntax is AT&T syntax, rather than the more popular Intel syntax. This is due to technical constraints; assembly parsing is provided by LLVM and its support for Intel syntax is buggy and not well tested.
>
> Some day Zig may have its own assembler. This would allow it to integrate more seamlessly into the language, as well as be compatible with the popular NASM syntax. This documentation section will be updated before 1.0.0 is released, with a conclusive statement about the status of AT&T vs Intel/NASM syntax. 

Sweet, they'll add NASM syntax!